### PR TITLE
Tags in Projects

### LINKED ISSUE #182: Tags in projects
#### I am submitting a -
* [x] **Feature Request**
* [] **Bug Report**

#### Describe your bug/feature -
Tags in projects.
#### Current behaviour -
Currently there are no tags in projects.
#### Expected behaviour -
We can have an option to tag projects and with the help of these tags we can view all projects by opening tags/<any_tag> also this can be used to generate queries to a very primtive search engine which will act as a placehoder until a good search engine is implemented.   
#### Are you working on this issue? (Yes/No)
I am working on implementing tags right now.

### PR DESCRIPTION
**Pull request by shubhankarsharma00**
Fixes #182 

#### Describe the changes you have made in this pr -
I have added a very primitive way to add tags and make a query based on tags which can be built upon to make a very primitive search engine which can be used as a placeholder until a good search engine is implemented.

### CODE REVIEW

**Comment by shubhankarsharma00** [comment]
@satu0king or @tachyons I tried to make a demo model can you please review this and suggest changes.

File: app/models/tag.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class Tag < ApplicationRecord
+	has_many :taggings
+	has_many :projects, through: :taggings
```
Comment (tachyons) [inline app/models/tag.rb]: indentation

File: db/migrate/20190303182743_create_taggings.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class CreateTaggings < ActiveRecord::Migration[5.1]
+  def change
+    create_table :taggings do |t|
+      t.belongs_to :project, foreign_key: true
+      t.belongs_to :tag, foreign_key: true
+
```
Comment (ankitkataria) [inline db/migrate/20190303182743_create_taggings.rb]: Is it possible to add indexes here and also in the tags table? like [so](https://stackoverflow.com/questions/223804/indexes-for-has-many-through). This'll help with the queries.

Comment (shubhankarsharma00) [reply db/migrate/20190303182743_create_taggings.rb]: @ankitkataria  this was just a primitive model we can add indexes.

File: app/models/tag.rb (hunk 1 above)
Comment (shubhankarsharma00) [reply app/models/tag.rb]: @tachyons I probably missed it cause the file was autogenerated. I have fixed it now.

**Comment by jbox144** [comment]
image
11/10 would project again

**Comment by shubhankarsharma00** [comment]
@jbox144 sorry can't undestand what you're trying to say.

**Comment by shubhankarsharma00** [comment]
Right now I'm thinking of adding indexes and I'll work a bit on it's frontend. Anything else I should focus on to improve this feature?

**Comment by tachyons** [comment]
Rebase with master and add tests for newly added models

**Comment by tachyons** [comment]
Also remove auto generated files which are not needed

**Comment by shubhankarsharma00** [comment]
@tachyons okay 👍

**Comment by satu0king** [comment]
@shubhankarsharma00 try and rebase your branch with circuitverse/master as it is difficult to review with multiple changes.

**Comment by satu0king** [comment]
Can you style the tags so that they come in a single line in the project page? 
image

**Comment by shubhankarsharma00** [comment]
Yeah, Also I'm thinking of suggesting already existing tags to the user so the user can tag stuff easily.

**Comment by shubhankarsharma00** [comment]
@armudgal are these the changes you desired?
PS: @garry329 thanks for the search page

**Comment by shubhankarsharma00** [comment]
@tachyons I have added tests for the new models can you please review them.

File: app/views/projects/get_projects.html.erb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+<h2>Searched Tags: <%= params[:tag]%></h2>
+<% @projects.each do |project| %>
+  <div class="row feature m-4">
+  <div class="col-md-9">
+  <h3 class="card-title"><%= project.name %></h3>
+  <div class="description"><%= project.description %></div>
+  <div class="row mt-3">
+  <div class="col-4">
+            <div class="mt-3 mb-3">
+              <a class="btn btn-dark btn-md" target="_blank" href="<%= user_project_path(project.author, project) %>" role="button">View</a>
```
Comment (tachyons) [inline app/views/projects/get_projects.html.erb]: Indent html

File: public/js/embedListeners.js [hunk 4]
```diff
@@ -96,6 +96,21 @@ function startListeners() {
             }
 
         }
+        if (simulationArea.lastSelected && simulationArea.lastSelected.keyDown2) {
```
Comment (tachyons) [inline public/js/embedListeners.js]: related to this PR ?

**Comment by tachyons** [comment]
1. Rebase with master and make CI 💚 
2. Remove auto generated `test/` files
3. Squash commits
4. :shipit:

File: public/js/embedListeners.js (hunk 4 above)
Comment (shubhankarsharma00) [reply public/js/embedListeners.js]: This was a pr that was merged earlier.

**Comment by tachyons** [comment]
4.  https://www.quora.com/On-GitHub-what-is-the-significance-of-the-Ship-It-squirrel/answer/Joel-Glovier

**Comment by shubhankarsharma00** [comment]
@tachyons I'm not sure about these tests never used rake tests before.

File: app/views/projects/get_projects.html.erb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        <div class="description">Tags:
+          <%= raw project.tags.map(&:name).map { |t| link_to t, tag_path(t),{:style=>"background-color: #f1f1f1; color: #000000", :class=>"badge"} }.join(', ') %>
+        </div>
+      </div>
+      <div class="col-md-3">
+        <img src="<%= project.image_preview.url %>" alt="<%= project.name %>" class="img-fluid img-thumbnail">
+      </div>
+    </div>
+  <% end %>
+</div>
```
Comment (armudgal) [inline app/views/projects/get_projects.html.erb]: Can we add a blank line at the end. No linting reasons, it just looks better. Also no ugly messages such as `No newline at the end`

File: spec/models/tag_spec.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+require 'rails_helper'
+
+RSpec.describe Tag, type: :model do
+  describe "associations" do
+    it { should have_many(:taggings) }
+    it { should have_many(:projects) }
+  end
+end
```
Comment (armudgal) [inline spec/models/tag_spec.rb]: Blank line

File: spec/models/tagging_spec.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+require 'rails_helper'
+
+RSpec.describe Tagging, type: :model do
+  describe "associations" do
+    it { should belong_to(:project) }
+    it { should belong_to(:tag) }
+  end
+end
```
Comment (armudgal) [inline spec/models/tagging_spec.rb]: Blank line.

**Comment by armudgal** [comment]
@tachyons I am happy that I am learning new things everyday

**Comment by armudgal** [comment]
@tachyons Please review and :shipit:

**Comment by armudgal** [comment]
@shubhankarsharma00 Hey there, looks like one of the tests are failing. Can you please fix that and then we can go ahead with merging this.

**Comment by shubhankarsharma00** [comment]
@armudgal I was busy in a Hackathon I'll solve this today I guess rebasing with latest project tests would work.

File: app/controllers/projects_controller.rb [hunk 8]
```diff
@@ -12,7 +12,10 @@ class ProjectsController < ApplicationController
   def index
     @author = User.find(params[:user_id])
   end
-
+  # GET /projects/tags/[tag]
+  def get_projects
+    @projects = Project.tagged_with(params[:tag]).where(project_access_type: "Public")
```
Comment (tachyons) [inline app/controllers/projects_controller.rb]: Create a scope in project

```ruby
  class Project
    scope :public, -> { where(project_access_type: "Public") }
  end
```
```ruby
  @projects = Project.public.tagged_with(params[:tag]) 
```

File: app/models/project.rb [hunk 9]
```diff
@@ -8,7 +8,8 @@ class Project < ApplicationRecord
 
   has_many :collaborations, dependent: :destroy
   has_many :collaborators, source: 'user' , through: :collaborations
-
+  has_many :taggings
```
Comment (tachyons) [inline app/models/project.rb]: `dependent: :destroy`

**Comment by shubhankarsharma00** [comment]
@tachyons does it seem okay now?

File: app/models/project.rb [hunk 10]
```diff
@@ -8,15 +8,15 @@ class Project < ApplicationRecord
...
+  has_many :taggings, dependent: :destroy
   has_many :tags, through: :taggings
   mount_uploader :image_preview, ImagePreviewUploader
 
   self.per_page = 8
 
   acts_as_commontable
   # after_commit :send_mail, on: :create
-
+  scope :public_view, -> { where(project_access_type: "Public") }
```
Comment (tachyons) [inline app/models/project.rb]: just `public` is enough

File: app/models/project.rb [hunk 11]
```diff
@@ -65,19 +65,19 @@ def send_mail
   end
 
   def self.tagged_with(name)
-    Tag.find_by!(name: name).projects
+    Tag.find_by!(name: name).projects.public_view
```
Comment (tachyons) [inline app/models/project.rb]: Keep public scope where we need ie in the controller action, What if need to fetch private projects of a user with a particular tag in future ?.

File: app/models/project.rb (hunk 10 above)
Comment (shubhankarsharma00) [reply app/models/project.rb]: rails 4.1 doesn't allow the use of reserved words like public, private anymore.

File: app/models/project.rb (hunk 11 above)
Comment (shubhankarsharma00) [reply app/models/project.rb]: yeah makes sense I'll change this asap.

File: app/models/project.rb (hunk 10 above)
Comment (tachyons) [reply app/models/project.rb]: My bad!. Can we have better name then? What about `open` ?

Comment (shubhankarsharma00) [reply app/models/project.rb]: ``` open ``` seems nice :wink:

**Comment by satu0king** [comment]
Looks good to me, thanks!

**Pull request merged**